pkg/auth/auth.go

```diff
@@ -147,24 +170,65 @@ func Login(ctx context.Context, systemContext *types.SystemContext, opts *LoginO
 		return nil
 	}
 	if unauthorized, ok := err.(docker.ErrUnauthorizedForCredentials); ok {
-		logrus.Debugf("error logging into %q: %v", server, unauthorized)
-		return errors.Errorf("error logging into %q: invalid username/password", server)
+		logrus.Debugf("error logging into %q: %v", key, unauthorized)
+		return errors.Errorf("error logging into %q: invalid username/password", key)
 	}
-	return errors.Wrapf(err, "authenticating creds for %q", server)
+	return errors.Wrapf(err, "authenticating creds for %q", key)
+}
+
+// getCredentialsForRef returns the auth config for the provided key as well as its registry.
+func getCredentialsForRef(systemContext *types.SystemContext, key string) (auth types.DockerAuthConfig, registry string, err error) {
+	trimmedKey := trimScheme(key)
+	if trimmedKey != key {
+		return auth, registry, errors.New("credentials key has https[s]:// prefix")
+	}
+
+	// Non-canonical references have to be handled by using the legacy API.
+	domain := getRegistryName(trimmedKey)
+	if trimmedKey == domain {
+		// nolint: staticcheck
+		auth, err := config.GetCredentials(systemContext, domain)
+		return auth, domain, err
+	}
+
+	// Registry path provided, try to get the credentials for the ref.
+	ref, err := reference.ParseNamed(trimmedKey)
+	if err != nil {
+		return auth, registry, errors.Wrapf(err, "parse reference from %q", trimmedKey)
+	}
+
+	if !reference.IsNameOnly(ref) {
```

Comment:
This should happen in a parsing step separate from `[gG]etCredentials*`, so that the initial `RemoveAuthentication` also refuses to work on a tagged/digested reference.

`func parseRegistryArgument(arg string, acceptRepositories bool) (key string, registry string, refIfNamespaced *reference.Named)` or something like that?

(Arguably it’s actually `SetCredentials`/`RemoveAuthentication` where it would be more important to reject tagged/digested keys, so that the integrity of the file is preserved; and then this caller could afford to be a bit sloppy.)

Reply:
Good, added a new argument parse function as well as unit tests for it.